### PR TITLE
Clean restricted memory access in mm cpu

### PR DESCRIPTION
**Pull request by jiaxinwu**
# Description

<_Include a description of the change and why this change was made._>

<_For each item, place an "x" in between `[` and `]` if true. Example: `[x]` (you can also check items in GitHub UI)_>

<_Create the PR as a Draft PR if it is only created to run CI checks._>

<_Delete lines in \<\> tags before creating the PR._>

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

<_Describe the test(s) that were run to verify the changes._>

## Integration Instructions

<_Describe how these changes should be integrated. Use N/A if nothing is required._>

### CODE REVIEW

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:1]: Can you explain in detail that #PF only happens when SmiProfile is enabled so CR2 can be saved/restored in SmiProfile enable configuration?

And CR2 saving/restoring we think is not heavy operation as CR3 saving/restoring. So we remove the check of SmiProfile and always save/restore CR2.

**Comment by kraxel** [comment]
Not working in qemu.
```
!!!! X64 Exception Type - 0D(#GP - General Protection)  CPU Apic ID - 00000000 !!!!
ExceptionData - 0000000000000000
RIP  - 000000003FFC452C, CS  - 0000000000000038, RFLAGS - 0000000000010002
RAX  - 0000000000000001, RCX - 000000000000017D, RDX - 000000000F8BFBFF
RBX  - 0000000000000000, RSP - 000000003FFB8D48, RBP - 000000003FFB8DC0
RSI  - 0000000000000000, RDI - 0000000000000000
R8   - 0000000000000000, R9  - 0000000000000000, R10 - 0000000000000000
R11  - 0000000000000000, R12 - 0000000000000000, R13 - 0000000000000000
R14  - 00000000FEE00300, R15 - 0000000000000000
DS   - 0000000000000020, ES  - 0000000000000020, FS  - 0000000000000020
GS   - 0000000000000020, SS  - 0000000000000020
CR0  - 0000000080010033, CR2 - 0000000000000000, CR3 - 000000003FF83000
CR4  - 0000000000000668, CR8 - 0000000000000000
DR0  - 0000000000000000, DR1 - 0000000000000000, DR2 - 0000000000000000
DR3  - 0000000000000000, DR6 - 00000000FFFF0FF0, DR7 - 0000000000000400
GDTR - 000000003FFAE000 000000000000004F, LDTR - 0000000000000000
IDTR - 000000003FFB2000 00000000000001FF,   TR - 0000000000000040
FXSAVE_STATE - 000000003FFB89A0
!!!! Find image based on IP(0x3FFC452C) /home/kraxel/projects/edk2/Build/OvmfX64/DEBUG_GCC5/X64/UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuDxeSmm/DEBUG/PiSmmCpuDxeSmm.dll (ImageBase=000000003FFBA000, EntryPoint=000000003FFC9CA1) !!!!
```

File: UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/X64/PageTbl.c:1]: agree, updated commit message

**Comment by kraxel** [comment]
> build command: build -p OvmfPkg\OvmfPkgX64.dsc -a X64 -D SECURE_BOOT -D SMM_REQUIRE -D DEBUG_ON_SERIAL_PORT

Using `-D SMM_REQUIRE=TRUE -D SECURE_BOOT_ENABLE=TRUE` here.

> run qume command: qemu-system-x86_64 -machine q35,smm=on -m 4G -net none -drive if=pflash,format=raw,unit=0,file=OVMF_CODE.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=OVMF_VARS.fd -hda fat:rw:hda-contents -serial file:log.log

Additional arguments: `-cpu host -enable-kvm` (on linux host).

Tail of the linux debug console:
```
Loading SMM driver at 0x0003FFBA000 EntryPoint=0x0003FFC9CA1 PiSmmCpuDxeSmm.efi
mAddressEncMask = 0x0
FindSmramInfo: SMRR Base = 0x3F000000, SMRR Size = 0x1000000
PcdCpuSmmCodeAccessCheckEnable = 1
PcdControlFlowEnforcementPropertyMask = 0
SMRAM TileSize = 0x00002000 (0x00001000, 0x00001000)
CPU[000]  APIC ID=0000  SMBASE=3EFF9000  SaveState=3F008C00  Size=00000400
Stacks                   - 0x3FFB3000
mSmmStackSize            - 0x6000
PcdCpuSmmStackGuard      - 0x1
One Semaphore Size    = 0x40
Total Semaphores Size = 0x1C0
PhysicalAddressBits = 39, 5LPageTable = 0.
5LevelPaging Needed             - 0
1GPageTable Support             - 1
PhysicalAddressBits             - 39
Initialize IDT IST field for SMM Stack Guard
```

**Comment by kraxel** [comment]
Oh, master branch crashes too, so this is not something in this PR.

**Comment by kraxel** [comment]
bisect landed here: https://github.com/tianocore/edk2/commit/268397a8923

**Comment by jiaxinwu** [comment]
> bisect landed here: [268397a8923](https://github.com/tianocore/edk2/commit/268397a8923)

Thanks Gred, i guess it's related the MCA CAP register not supported the qemu part. So, provide the https://github.com/tianocore/edk2/pull/6166 to skip the MCA CAP register access if cpu doesn't support it. i will try to reproduce the issue on the linux host, meanwhile, could you also help check whether https://github.com/tianocore/edk2/pull/6166 can resolve the problem?

**Comment by jiaxinwu** [comment]
> -D SMM_REQUIRE=TRUE -D SECURE_BOOT_ENABLE=TRUE

I have root cause the issue on linux, and confirmed https://github.com/tianocore/edk2/pull/6166 can resolve the probelm.

**Review by niruiyu — APPROVED** [review]

**Pull request merged**